Implement new difficulty algorithm

str4d opened the pull request:
Closes #931. Closes #696.

str4d commented:
Requires regenerating `miner_tests`, which I will do when it is this PR's turn to be merged :)

defuse commented on src/chainparams.cpp:
```diff
@@ -39,7 +39,9 @@ class CMainParams : public CChainParams {
-        consensus.nPowTargetTimespan = 14 * 24 * 60 * 60; // two weeks
+        consensus.nPowMaxAdjustDown = 16; // 16% adjustment down
+        consensus.nPowMaxAdjustUp = 8; // 8% adjustment up
+        consensus.nPowTargetTimespan = 17 * 10 * 60; // 17 blocks
```
Could the "10" here be replaced by a constant with a name indicating it's the target time in minutes? Does such a constant already exist?

defuse replied:
Reason: We're going to change the target time and don't want this to be forgotten.

defuse commented on src/pow.cpp:
```diff
@@ -34,36 +32,37 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
-        nActualTimespan = params.nPowTargetTimespan*4;
+
+    int64_t nMinActualTimespan = params.nPowTargetTimespan * (100 - params.nPowMaxAdjustUp) / 100;
+    int64_t nMaxActualTimespan = params.nPowTargetTimespan * (100 + params.nPowMaxAdjustDown) / 100;
```
Maybe add braces around the multiplication so that it's obvious (without having to look up precedence/associativity rules) that the `/ 100` doesn't happen first resulting in `nMinActualTimespan = 0` and `nMaxActualTimespan = params.nPowTargetTimespan`.

defuse commented on src/chainparams.cpp:
```diff
@@ -39,7 +39,9 @@ class CMainParams : public CChainParams {
+        consensus.nPowMaxAdjustDown = 16; // 16% adjustment down
+        consensus.nPowMaxAdjustUp = 8; // 8% adjustment up
+        consensus.nPowTargetTimespan = 17 * 10 * 60; // 17 blocks
         consensus.nPowTargetSpacing = 10 * 60;
```
Oh never mind, the target spacing is right here. When we're changing this constant we should notice the one above :)

defuse commented:
ACK. Looks like it matches the behavior of DigiByte's implementation. I left a nitpick comment here: https://github.com/zcash/zcash/pull/1004#discussion_r66296132

str4d commented:
Rebased on latest, and adjusted the consensus parameters so we don't have two parameters encoding the same information (the target block interval).

defuse commented on src/rpcmining.cpp:
```diff
@@ -45,9 +45,9 @@ Value GetNetworkHashPS(int lookup, int height) {
+    // If lookup is -1, then use difficulty averaging window.
     if (lookup <= 0)
-        lookup = pb->nHeight % Params().GetConsensus().DifficultyAdjustmentInterval() + 1;
+        lookup = pb->nHeight - Params().GetConsensus().nPowAveragingWindow;
```
Might this still be negative if there are less than nPowAveragingWindow blocks?

str4d replied:
Good point!

str4d replied:
See just below here for the fix.

defuse commented on src/chainparams.cpp:
```diff
@@ -39,7 +39,9 @@ class CMainParams : public CChainParams {
-        consensus.nPowTargetTimespan = 14 * 24 * 60 * 60; // two weeks
+        consensus.nPowAveragingWindow = 17;
+        consensus.nPowMaxAdjustDown = 16; // 16% adjustment down
+        consensus.nPowMaxAdjustUp = 8; // 8% adjustment up
```
These should be added to the Testnet and Regtest params, I think (if the default value of 0 is intended then make it explicit).

str4d replied:
There shouldn't be a default value of 0; being undefined in testnet and regtest, they should take on these same parameters.

defuse replied:
Ah, I see!

str4d replied:
(Note: `nPowMaxAdjustDown` and `nPowMaxAdjustUp` *have* now been added to regtest with values of 0, in order to disable difficulty adjustment in regtest mode.)

defuse commented:
Added some more comments

str4d commented:
Rebased on latest, addressed @defuse's comments, and disabled difficulty changes in regtest mode (ie. back to same operating assumptions for regtest mode as before - no difficulty changes in the interval we are interested in). I'll need to rebase one last time after @ebfull is done with #968.

ebfull commented:
I'm going to do a review pass after this is rebased.

ebfull commented:
@defuse ACK'd this, time for me to review.

ebfull commented:
LGTM, good work!

ebfull commented:
I want to give @daira a chance to look at this before we `r+`.

str4d commented on src/pow.cpp:
```diff
@@ -34,54 +32,53 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
     }
 
-    // Go back by what we want to be 14 days worth of blocks
-    int nHeightFirst = pindexLast->nHeight - (params.DifficultyAdjustmentInterval()-1);
```
For reference, this is the line that would have needed to be fixed by #696 if we were not changing the algorithm.

ebfull commented:
Added "Closes #696" to the PR.

daira commented on src/pow.cpp:
```diff
@@ -34,54 +32,53 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
+    LogPrint("pow", "GetNextWorkRequired RETARGET\n");
+    LogPrint("pow", "params.AveragingWindowTimespan() = %d    nActualTimespan = %d\n", params.AveragingWindowTimespan(), nActualTimespan);
+    LogPrint("pow", "Before: %08x  %s\n", pindexLast->nBits, bnOld.ToString());
+    LogPrint("pow", "After:  %08x  %s\n", bnNew.GetCompact(), bnNew.ToString());
```
What's the difference between ``LogPrint`` and ``LogPrintf``? I would have thought these (and two more above) needed to be ``LogPrintf`` since they use formatted parameters.

daira replied:
Oh, the difference is the "pow" argument. That's confusing naming.

str4d replied:
Yeah. `LogPrintf` lines are always logged; `LogPrint` lines are only logged if the first argument is set in `-debug`, IE `-debug=pow` in this case.

daira commented on src/rpcmining.cpp:
```diff
@@ -45,9 +45,13 @@ Value GetNetworkHashPS(int lookup, int height) {
         return 0;
 
-    // If lookup is -1, then use blocks since last difficulty change.
+    // If lookup is -1, then use difficulty averaging window.
```
Comment is inconsistent with code (``== -1`` vs ``<= 0``).

str4d replied:
I left it consistent with upstream, but can change (won't be for eight hours though).

daira commented on src/test/miner_tests.cpp:
```diff
@@ -158,7 +158,8 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
         pblock->nVersion = 4;
-        pblock->nTime = chainActive.Tip()->GetMedianTimePast()+1;
+        // Fake the blocks taking at least nPowTargetSpacing
+        pblock->nTime = chainActive.Tip()->GetMedianTimePast()+6*Params().GetConsensus().nPowTargetSpacing;
```
Where does the 6 come from? - can the comment explain that?

daira commented on src/test/pow_tests.cpp:
```diff
@@ -19,12 +19,12 @@ BOOST_AUTO_TEST_CASE(get_next_work)
+    int64_t nLastRetargetTime = 1262149169; // NOTE: Not an actual block time
     CBlockIndex pindexLast;
     pindexLast.nHeight = 32255;
     pindexLast.nTime = 1262152739;  // Block #32255
```
Should this comment also be changed?

daira replied:
And similarly twice below.

daira commented:
ACK modulo @defuse's comment https://github.com/zcash/zcash/pull/1004/files#r66839688 . I had some comment nits but they do not block.

defuse commented:
ACK for the fixes

daira commented on src/rpcmining.cpp:
```diff
@@ -45,9 +45,13 @@ Value GetNetworkHashPS(int lookup, int height) {
+
+    // If lookup is still negative, then use blocks since genesis.
+    if (lookup <= 0)
+        lookup = pb->nHeight;
```
Shouldn't this be ``if (lookup < 0) lookup = 0;``? I.e. of there are not enough blocks for the full window, we should still use however many blocks are available.

defuse replied:
Oh yeah, I think you're right.

daira replied:
... in which case using ``max(0, ...)`` would be simpler.

str4d replied:
No, because `lookup` is used as a counter, not an index into the blockchain. I'll merge this `if` with the existing one below it to make this clearer.

daira replied:
Oh, in that case what's confusing me is line 50. Why is that not ``lookup = Params().GetConsensus().nPowAveragingWindow;``?

str4d replied:
Oh! Whoops, you're right! I'll open a new PR to fix this.

str4d commented:
Comments addressed. Hopefully this is now ready for merging!

ebfull commented:
ACK

ebfull commented:
@zkbot r+

ebfull commented:
Great work on this @str4d.

daira commented:
ACK for the comment changes, but still confused by https://github.com/zcash/zcash/pull/1004#discussion_r67304866

str4d commented:
See #1030 for the fix to the above issue.
